ms_active_directory/core/ad_session.py

```diff
@@ -1934,6 +1956,56 @@ def find_primary_group_for_account(self, account: Union[str, ADUser, ADComputer]
+                                           fails.
+        :param skip_validation: If true, assume all distinguished names exist and do not look them up.
+                                Defaults to False. This can be used to make this function more performant when
+                                the caller knows all the distinguished names being specified are valid, as it
+                                performs far fewer queries.
+        :returns: True if the operation succeeds, False otherwise.
+        :raises: InvalidLdapParameterException if the account is not a string, ADUser, or ADComputer.
+        :raises: ObjectNotFoundException if the account cannot be found.
+        :raises: MembershipModificationException if we fail to add account to specified primary_group.
+        :raises: AttributeModificationException if raise_exception_on_failure is True and we fail
+        """
+        if isinstance(primary_group, ADGroup):
+            # if our group is an ADGroup, we can do the lookup by DN which is most efficient
+            informed_group_object = self.find_group_by_distinguished_name(primary_group.distinguished_name,
+                                                                          [ldap_constants.AD_ATTRIBUTE_OBJECT_SID],
+                                                                          controls=controls)
+        elif isinstance(primary_group, str):
+            informed_group_object = self.find_group_by_name(primary_group, [ldap_constants.AD_ATTRIBUTE_OBJECT_SID],
+                                                            controls=controls)
+        else:
+            raise InvalidLdapParameterException('primary_group must be either a string or an ADGroup object. {} is '
+                                                'neither. '.format(primary_group))
+        if informed_group_object is None:
+            raise ObjectNotFoundException('No group could be found in the domain with the Group object class using '
+                                          'group {}'.format(primary_group))
+        # get objectSid for the primary_group
+        group_sid = informed_group_object.get(ldap_constants.AD_ATTRIBUTE_OBJECT_SID)
+        group_rid = ldap_utils.get_rid_from_object_sid(group_sid)
+
+        account = self._validate_user_and_get_user_obj(account, can_be_computer=True, skip_validation=skip_validation)
+        # ensure the account is a member of the primary group before setting primaryGroupID attribute
+        # If the group add fails, it will throw a MembershipModificationException.
+        self.add_users_to_groups([account], [informed_group_object], controls=controls,
+                                 skip_validation=skip_validation)
+        return self.overwrite_attribute_for_object(account, ldap_constants.AD_ATTRIBUTE_PRIMARY_GROUP_ID, group_rid,
+                                                   controls, raise_exception_on_failure, skip_validation)
```

Comment:
is setting this as a string the right format? have you tested it?
SIDs are usually bytestrings, so I'm not sure if setting this as a string is necessarily the right format

Reply:
This actually sets it as an integer:
```
def get_rid_from_object_sid(object_sid: str) -> int:
    """ Given an objectSid, return the RID (Relative Identifier) of the object.
    """
    rid_string = object_sid[object_sid.rfind('-') + 1:]
    return int(rid_string)
```
According to [ldap wiki](https://ldapwiki.com/wiki/PrimaryGroupID), the syntax for `primaryGroupID` is an integer. I have tested this as well and it does work in my test cases.

Reply:
ah woops, I missed the conversion to int. thanks for clarifying.